### PR TITLE
Fix estimate_parameters for binned and non uniform axis

### LINKED ISSUE #2739: nan and maps non supported by current implementation of scaling factor for binned non-uniform axes
The new implementation of scaling factor for binned non-uniform #2728 axes created some issues. Eg in the gaussian implementation:

https://github.com/hyperspy/hyperspy/blob/d1bcd1c540c9a7489a5527f32b15b5680e12f55b/hyperspy/_components/gaussian.py#L165-L168

the evaluation of `value2index(centre)` at line 168 relies on the fact that `_estimate_gaussian_parameters ` at line 165 will output a non-NaN value for centre. 

```python
s = hs.signals.Signal1D(np.zeros(100))

nuax = FunctionalDataAxis(expression="x**2",x=s.axes_manager.signal_axes[0],name='NUAX')
snu = s.deepcopy()
snu.axes_manager.set_axis(nuax,index_in_axes_manager=0)

g = hs.model.components1D.Gaussian(A=1,centre=80,sigma=10)
g.estimate_parameters(snu,20.,900.)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-51-6b634f634055> in <module>
----> 1 g.estimate_parameters(snu,20.,900.)

c:\users\xxxxxxx\documents\git\hyperspy\hyperspy\_components\gaussian.py in estimate_parameters(self, signal, x1, x2, only_current)
    166                                                               only_current)
    167         scaling_factor = axis.scale if axis.is_uniform \
--> 168                          else np.gradient(axis.axis)[axis.value2index(centre)]
    169         if only_current is True:
    170             self.centre.value = centre

c:\users\xxxxxxx\documents\git\hyperspy\hyperspy\axes.py in value2index(self, value, rounding)
    567             return (np.abs(self.axis - value)).argmin()
    568         else:
--> 569             index = int(value)
    570             if self.size > index >= 0:
    571                 return index

ValueError: cannot convert float NaN to integer
```
This, regardless of whether the axis is binned or not

Additionnaly, this implementation also fails for signals with navigation dimensions > 0.  It is because `value2index(centre)` at line 168 relies on the fact that `_estimate_gaussian_parameters ` at line 165 will not output an array of values, which it will.

```python 
s = hs.signals.Signal1D(np.zeros((10,10,100)))

nuax = FunctionalDataAxis(expression="x**2",x=s.axes_manager.signal_axes[0],name='NUAX', bin)
snu = s.deepcopy()
snu.axes_manager.set_axis(nuax,index_in_axes_manager=2)

g = hs.model.components1D.Gaussian()

g.estimate_parameters(s,20.,40.)
True

g.estimate_parameters(snu,20.,400.)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-69-413522c6418a> in <module>
----> 1 g.estimate_parameters(snu,20.,40.)

c:\users\xxxxxx\documents\git\hyperspy\hyperspy\_components\gaussian.py in estimate_parameters(self, signal, x1, x2, only_current)
    166                                                               only_current)
    167         scaling_factor = axis.scale if axis.is_uniform \
--> 168                          else np.gradient(axis.axis)[axis.value2index(centre)]
    169         if only_current is True:
    170             self.centre.value = centre

c:\users\xxxxxx\documents\git\hyperspy\hyperspy\axes.py in value2index(self, value, rounding)
    564         if value is None:
    565             return None
--> 566         if self.low_value <= value <= self.high_value:
    567             return (np.abs(self.axis - value)).argmin()
    568         else:

ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```
Also regardless if the axes is binned or not.

I propose to fix that by:
-Ensuring that the computation of scaling_factor is performed only after checking that the axis is binned
-Ensuring that `centre` values are passed one-by-one to value2index
-Ensuring that  `NaN` values are forwarded by the function without disturbing its operation

 I will try to implement a patch asap

### PR DESCRIPTION
**Pull request by LMSC-NTappy**
### Description of the change
Rebase of PR #2740 with the correct version of `non_uniform_axes`.
Fixes #2739.

Summary: this PR fixes incorrect behaviour in `BaseDataAxis.value2index()` which was not accepting arrays as input and was computing the axis gradient even for non-binned axes.

### Progress of the PR
- [x] Change implemented (can be split into several points),
- [x] update docstring (if appropriate),
- [x] add tests,
- [x] ready for review.

### CODE REVIEW

File: hyperspy/axes.py [hunk 1]
```diff
@@ -559,22 +559,33 @@ def value2index(self, value, rounding=round):
...
         """
         if value is None:
             return None
-        if self.low_value <= value <= self.high_value:
-            return (np.abs(self.axis - value)).argmin()
+        #Should evaluate on both arrays and scalars. Raises error if there are
+        #nan values in array
+        if np.all((value >= self.low_value)*(value <= self.high_value)):
+            #initialise the index same dimension as input, force type to int
+            index = np.asarray(value).astype(int).copy()
```
Comment (ericpre) [inline hyperspy/axes.py]:
```suggestion
            index = np.empty_like(value, dtype=int)
```

File: hyperspy/tests/axes/test_data_axis.py [hunk 2]
```diff
@@ -314,6 +314,46 @@ def test_calibrate(self):
         with pytest.raises(TypeError, match="only for uniform axes"):
             self.axis.calibrate(value_tuple=(11,12), index_tuple=(0,5))
 
+    def test_functional_value2index(self):
```
Comment (ericpre) [inline hyperspy/tests/axes/test_data_axis.py:323]: Because this test is for a method of `BaseDataAxis`, it would be better to move this test to the `TestBaseDataAxis` class.

File: hyperspy/component.py [hunk 3]
```diff
@@ -1216,3 +1216,40 @@ def print_current_values(self, only_free=False, fancy=True):
...
+
+    """
+
+    if is_binned(signal):
+    # in v2 replace by
+    #if axis.is_binned:
+        if axis.is_uniform:
+            scaling_factor = axis.scale
+        else:
+            # if isinstance(parameter, (int, float)):
```
Comment (ericpre) [inline hyperspy/component.py]: Now that `value2index` takes array as input, this can be clean up.

File: hyperspy/axes.py [hunk 4]
```diff
@@ -559,22 +559,33 @@ def value2index(self, value, rounding=round):
...
-                    f'The value {value} is out of the limits '
-                    f'[{self.low_value:.3g}-{self.high_value:.3g}] of the '
-                    f'"{self._get_name()}" axis.'
-                    )
+            ### I commented out this because it seemed dangerous
+            # index = int(value)
+            # if self.size > index >= 0:
+            #     return index
+            # else:
+            raise ValueError(
```
Comment (ericpre) [inline hyperspy/axes.py:608]: Add a test for this case?

File: hyperspy/axes.py [hunk 5]
```diff
@@ -559,22 +559,33 @@ def value2index(self, value, rounding=round):
...
-            if self.size > index >= 0:
-                return index
-            else:
-                raise ValueError(
-                    f'The value {value} is out of the limits '
-                    f'[{self.low_value:.3g}-{self.high_value:.3g}] of the '
-                    f'"{self._get_name()}" axis.'
-                    )
+            ### I commented out this because it seemed dangerous
+            # index = int(value)
```
Comment (ericpre) [inline hyperspy/axes.py]: Yes, I agree and I ignore it because, I think that it may be better to discuss this in a separate PR. If you want to fix this in this PR, the `value2index` of `UniformDataAxis` needs to be changed consistently.

File: hyperspy/tests/axes/test_data_axis.py [hunk 6]
```diff
@@ -314,6 +314,46 @@ def test_calibrate(self):
...
+        arval[1,1] = -0.3
+        with pytest.raises(ValueError):
+            self.axis.value2index(arval)
+        #One NaN in array in --> error out
+        arval[1,1] = np.nan
+        with pytest.raises(ValueError):
+            self.axis.value2index(arval)
+        #Single-value-array-in --> scalar out
+        arval = np.array([1.0])
+        assert type(self.axis.value2index(arval)) == np.int32
```
Comment (ericpre) [inline hyperspy/tests/axes/test_data_axis.py]: This test is failing:
```python
        arval = np.array([1.0])
>       assert type(self.axis.value2index(arval)) == np.int32
E       AssertionError: assert <class 'numpy.int64'> == <class 'numpy.int32'>
E        +  where <class 'numpy.int64'> = type(1)
```

maybe use something along the line of:
```python
np.issubdtype(self.axis.value2index(arval), int)
```

**Review by ericpre — COMMENTED** [review]
This looks good to me and I have left a couple of comments.
I will leave someone else to review/merge it since I influenced this PR!

File: hyperspy/tests/axes/test_data_axis.py (hunk 2 above)
Comment (LMSC-NTappy) [reply hyperspy/tests/axes/test_data_axis.py:323]: I thought too, but since BaseDataAxis can practically not be instantiated it felt awkward. It is, indeed, a method of BaseDataAxis but will never ever be called outside the subclasses. 

I can still move it there if you think that's better. I'll need to do something like 
```python
bax = BaseDataAxis()
bax.size = 100
etc...
```

File: hyperspy/component.py (hunk 3 above)
Comment (LMSC-NTappy) [reply hyperspy/component.py]: Sure, I just forgot

File: hyperspy/axes.py (hunk 4 above)
Comment (LMSC-NTappy) [reply hyperspy/axes.py:608]: There are, it's the assert that cover nan input and out-of-bound values

File: hyperspy/axes.py (hunk 1 above)
Comment (LMSC-NTappy) [reply hyperspy/axes.py]: Ok, you are right it's way more elegant

File: hyperspy/tests/axes/test_data_axis.py (hunk 6 above)
Comment (LMSC-NTappy) [reply hyperspy/tests/axes/test_data_axis.py]: By the time i saw your comments I had already changed it to `np.isscalar` which I think works too but I'll change it to `np.issubdtype` if you think it's best

File: hyperspy/axes.py (hunk 5 above)
Comment (LMSC-NTappy) [reply hyperspy/axes.py]: This change already brings consistency with UniformDataAxis.

As implemented, FunctionalDataAxis.value2index allows for something very wrong to happen, ie it interpretates the value as an index if it can interpretate it as an int.

_Voyez plutot_

```python
>>> from hyperspy.axes import FunctionalDataAxis,UniformDataAxis
>>> uax = UniformDataAxis(size=101.0,offset=0.0,scale=0.01)
>>> uax.value2index(50)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-17-5cf2ab8da91f> in <module>
      1 uax = UniformDataAxis(size=101.0,offset=0.0,scale=0.01)
----> 2 uax.value2index(50)

c:\users\nicolastappy\documents\git\hyperspy\hyperspy\axes.py in value2index(self, value, rounding)
   1177                 return index
   1178             else:
-> 1179                 raise ValueError("The value is out of the axis limits")
   1180 
   1181     def update_axis(self):

ValueError: The value is out of the axis limits
```

Which is expected. Now
```python
>>> nuax = FunctionalDataAxis(expression="x**2",x=uax)
>>> nuax.value2index(69.69)
69
>>> nuax.value2index(99)
99
```
Which is absolutely wrong since `nuax` spans 0 to 1 like `uax`. This works with `uax` because double values are converted to indices before. I think this was placed here with the intent to let value2index fall back on `int==index`
like `isig[]` slices but in fact, it's simply wrong.

I am fine making another PR about this, but I'd rather have this fixed ASAP. If everyone is fine with that I'll just like to delete these line that are commented out

**Comment by jlaehne** [comment]
Looks good on first sight, I'll have a closer look in the next days when I find some time.

File: hyperspy/axes.py (hunk 4 above)
Comment (ericpre) [reply hyperspy/axes.py:608]: Yes, indeed, sorry, I may have confused myself with the previous PR, when copying and pasting the review which I started in #2740!

File: hyperspy/tests/axes/test_data_axis.py (hunk 6 above)
Comment (ericpre) [reply hyperspy/tests/axes/test_data_axis.py]: Yes, `np.isscalar` is fine too!

File: hyperspy/axes.py (hunk 5 above)
Comment (ericpre) [reply hyperspy/axes.py]: I looked at it in more depth, and indeed this is no reason for this to be there. When looking at it quickly, I used the `UniformDataAxis.value2index` as a reference, but it is different, because the index is calculated before. I was expecting this to follow a logic but I looks to me that when this was implemented, this fell through the cracks!

I would suggest to remove it because the if block before that cover scalar and arrays. Would it be possible to remove `UniformDataAxis.value2index` and use `BaseDataAxis.value2index` instead?

File: hyperspy/tests/axes/test_data_axis.py (hunk 2 above)
Comment (ericpre) [reply hyperspy/tests/axes/test_data_axis.py:323]: I would say that it would still be more tidy to have this test in `TestBaseDataAxis` but at the same time, this is true that most test are in `TestDataAxis`!
I was suggesting this because it happen to me in the past to write tests which were already written somewhere else, just not where I expected them to be...

**Review by ericpre — APPROVED** [review]
This looks good to me, just need a tiny bit of tidying up. @jlaehne, I leave it to you. Thanks!

File: hyperspy/axes.py (hunk 5 above)
Comment (LMSC-NTappy) [reply hyperspy/axes.py]: I will do that then.

I will then also implement the "rounding" argument (which I actually forgot in the first place)

Comment (LMSC-NTappy) [reply hyperspy/axes.py]: > I would suggest to remove it because the if block before that cover scalar and arrays. Would it be possible to remove `UniformDataAxis.value2index` and use `BaseDataAxis.value2index` instead?

This failed pretty badly. That requires to support: string input with units, and relative position in the axis. So  `UniformDataAxis.value2index` uses some of `UnitConversion` methods to parse string inputs. I think UnitConversion should be made compatible with FunctionalDataAxis in another PR. 

I forced push back my latest modifications

Comment (ericpre) [reply hyperspy/axes.py]: Agree, let's finish this PR and #2747 and then it would be easier to remove `UniformDataAxis.value2index`.

File: hyperspy/_components/exponential.py [hunk 7]
```diff
@@ -52,7 +52,7 @@ class Exponential(Expression):
     """
 
     def __init__(self, A=1., tau=1., module="numexpr", **kwargs):
-        super(Exponential, self).__init__(
+        super().__init__(
```
Comment (jlaehne) [inline hyperspy/_components/exponential.py:55]: Why remove the attributes here and not for other components?

File: hyperspy/signal.py [hunk 8]
```diff
@@ -2656,7 +2656,13 @@ def get_1D_sum_explorer_wrapper(*args, **kwargs):
             navigator = self
             # Sum over all but the first navigation axis.
             am = navigator.axes_manager
-            navigator = navigator.sum(am.signal_axes + am.navigation_axes[1:])
+            with warnings.catch_warnings():
+                warnings.filterwarnings("ignore", category=UserWarning,
+                                        module='hyperspy'
+                                        )
+                navigator = navigator.sum(
```
Comment (jlaehne) [inline hyperspy/signal.py:2663]: It would make sense to use integrate1D here, which correctly decides on whether a `sum` can be made (binned signal) or an integration is needed (esp. non-uniform, unbinned signal)  -- in fact, it is on the todo-list in #2398. Then we could avoid catching the warning here.

File: hyperspy/signal.py [hunk 9]
```diff
@@ -4317,7 +4323,11 @@ def integrate1D(self, axis, out=None):
         if is_binned(self, axis=axis):
         # in v2 replace by
         # self.axes_manager[axis].is_binned
-            return self.sum(axis=axis, out=out)
+            with warnings.catch_warnings():
```
Comment (jlaehne) [inline hyperspy/signal.py]: Isn't this catch obsolete as this warning should be raised only if `is_binned = False`: https://github.com/hyperspy/hyperspy/blob/914bebd2944f3d63d9ffe02a52aa11d0f311e5a9/hyperspy/signal.py#L3700

File: hyperspy/tests/axes/test_data_axis.py (hunk 2 above)
Comment (jlaehne) [reply hyperspy/tests/axes/test_data_axis.py:323]: Couldn't we add a comment to `TestBaseDataAxis` to indicate that this is tested in `TestDataAxis` @ericpre ?

Comment (jlaehne) [reply hyperspy/tests/axes/test_data_axis.py:323]: Actually, it is checked for all three axes types individually, so that could be mentioned in the `TestBaseDataAxis` class.

Comment (jlaehne) [inline hyperspy/tests/axes/test_data_axis.py:323]: For the other two axes types, the different types of inputs for `value2index` and possible errors are tested in a number of individual functions, while here they are all in one function. Wouldn't it make sense to decide for one way for all three axes types? I'm fine with either.

Comment (jlaehne) [reply hyperspy/tests/axes/test_data_axis.py:323]: Thinking about it, the single function is probably more clear. Would you want to adapt it for the other axes types?

**Review by jlaehne — COMMENTED** [review]
Thanks a lot for the help with the nua branch! I left a few questions, but as said previously this looks good otherwise.

File: hyperspy/axes.py (hunk 5 above)
Comment (jlaehne) [reply hyperspy/axes.py]: Add to todo list in #2398?

File: hyperspy/_components/exponential.py (hunk 7 above)
Comment (ericpre) [reply hyperspy/_components/exponential.py:55]: This is old python 2 syntax (https://www.python.org/dev/peps/pep-3135, https://docs.python.org/3/library/functions.html#super), when I come across them, I trend to remove them...

File: hyperspy/tests/axes/test_data_axis.py (hunk 2 above)
Comment (LMSC-NTappy) [reply hyperspy/tests/axes/test_data_axis.py:323]: >Actually, it is checked for all three axes types individually, so that could be mentioned in the TestBaseDataAxis class.

Ok i will do it. 

I would suggest, in a later PR, to formalize the fact that BaseDataAxis is an abstract class, and provide guidelines for the implementation of different axes types.

File: hyperspy/axes.py (hunk 5 above)
Comment (LMSC-NTappy) [reply hyperspy/axes.py]: On a second thought, I think that keeping the possibility of overriding `BaseDataAxis.value2index` is still a good option. Coding universal axes methods in `BaseDataAxis` is smart, because it allows users to easily adapt hyperspy to their needs, while relying on basic functionality. 

However, I would argue that a few specific cases (`UniformDataAxis `being the most common one) will cover most of the experimental needs. For those, keeping override of `BaseDataAxis ` as a standard encourages the implementation of optimised code in dedicated axis classes 

Here is a picture of `UniformDataAxis.value2index()` vs `BasedataAxis.value2index()`, which, I think, is self-explanatory
 
image

File: hyperspy/_components/exponential.py (hunk 7 above)
Comment (LMSC-NTappy) [reply hyperspy/_components/exponential.py:55]: I can take care of removing it from other components then

File: hyperspy/axes.py (hunk 5 above)
Comment (ericpre) [reply hyperspy/axes.py]: Do you know which part is slow? If so, can it be fixed?

File: hyperspy/signal.py (hunk 8 above)
Comment (ericpre) [reply hyperspy/signal.py:2663]: Is there a reason to use `integrate1D` in favour of `sum`?

Using `integrate_simpson` will not work for `len(am.signal_axes + am.navigation_axes[1:] > 1` where `am` is the `axes_manager`.

The following example will not work:
```python
import hyperspy.api as hs
import numpy as np

s = hs.signals.Signal2D(np.arange(5*10*15).reshape(5, 10, 15))
s.plot(navigator='spectrum')
```

Comment (LMSC-NTappy) [reply hyperspy/signal.py:2663]: The idea is to preserve the physical meaning of the signal I presume. But I agree that this is out of reach for a simple PR like this one.

The objective of this pull request is just to fix some faulty behaviour in `BaseDataAxis.value2index()` (I updated the description to make this clear) but I am fine with attacking the `Integrate1D` question in another PR

File: hyperspy/_components/exponential.py (hunk 7 above)
Comment (LMSC-NTappy) [reply hyperspy/_components/exponential.py:55]: ...Done

File: hyperspy/axes.py (hunk 5 above)
Comment (LMSC-NTappy) [reply hyperspy/axes.py]: It cannot unfortunately. `np.argmin(axis-value)` is the nominal strategy for finding a closest value using numpy, as far as I know.
Still it is O(N). In comparison, explicit inversion of the axis function as in `UniformDataAxis` is probably O(1) or something close. 

I don't see any way to improve the speed of this. This needs to be applicable in `DataAxis`, which is purely value-defined.

A solution can be found in `FunctionalDataAxis` though, provided the function can be inverted.

File: hyperspy/signal.py (hunk 8 above)
Comment (jlaehne) [reply hyperspy/signal.py:2663]: Yes, that was the idea - but probably you're right to leave this one for later.

File: hyperspy/signal.py (hunk 9 above)
Comment (jlaehne) [reply hyperspy/signal.py]: Only open point remaining.

Comment (LMSC-NTappy) [reply hyperspy/signal.py]: I think this is necessary until the v2 because `is_binned` will evaluate from the signal attribute until then. So `is_binned` could evaluate `True` and then `not ax.is_binned` could evaluate `True` as well, if you integrate on a non-uniform-non-binned axis of a signal that otherwise has `signal.metadata.Signal.binned = True` set due to another axis. 

See here:
https://github.com/hyperspy/hyperspy/blob/0c6da6ec9ca1d27b6bc33a3f5e4ad31339bcb45c/hyperspy/misc/utils.py#L1430-L1438

(I did not implement this change however )

File: hyperspy/tests/axes/test_data_axis.py (hunk 2 above)
Comment (LMSC-NTappy) [reply hyperspy/tests/axes/test_data_axis.py:323]: Should be good now

File: hyperspy/signal.py (hunk 9 above)
Comment (ericpre) [reply hyperspy/signal.py]: Thanks, your are both correct and this is now fixed in 2e757e9.

File: hyperspy/axes.py (hunk 5 above)
Comment (ericpre) [reply hyperspy/axes.py]: Assuming the for loop is the cause of the poor performance, could it be vectorised? If not, use numba?

Comment (LMSC-NTappy) [reply hyperspy/axes.py]: I agree to try.

I will try to have numba'ed functions for rounding and checking if values are within limits. numba_closest_index_round, numba_closest_index_floor, numba_closest_index_ceil, numba_check_limits and let the BaseDataAxis.value2index take care of input formatting

File: hyperspy/axes.py [hunk 10]
```diff
@@ -559,22 +559,50 @@ def value2index(self, value, rounding=round):
...
+            #initialise the index same dimension as input, force type to int
+            index = np.empty_like(value,dtype=int)
+            #assign on flat, iterate on flat.
+            for i,v in enumerate(np.asarray(value).flat):
+                index.flat[i] = rounding_index(self.axis - v)
+            #Squeezing to get a scalar out if scalar in. See squeeze doc
+            return np.squeeze(index)[()]
+        else:
+            raise ValueError(
+                f'The value {value} is out of the limits '
```
Comment (jlaehne) [inline hyperspy/axes.py:609]: Can you add a test for this line (see codecov warning)?

File: hyperspy/axes.py [hunk 11]
```diff
@@ -559,22 +559,50 @@ def value2index(self, value, rounding=round):
         ------
         ValueError
```
Comment (jlaehne) [inline hyperspy/axes.py:564]: Can you harmonize the docstrings of the two value2index functions (also Parameters section and first line) - or define a common string that is used in both.

File: hyperspy/axes.py (hunk 5 above)
Comment (LMSC-NTappy) [reply hyperspy/axes.py]: Ok it kinda works with numba.

However the argument stays the same. I could make it ~10 times faster with numba, I can probably make `UniformDataAxis.value2index() `10times faster as well

image

File: hyperspy/axes.py (hunk 10 above)
Comment (LMSC-NTappy) [reply hyperspy/axes.py:609]: Looks weird to me. This is the case where a value is out of axis and is tested here:

https://github.com/LMSC-NTappy/hyperspy/blob/2e757e940dd11b0d913cb1c78348beb54c133f63/hyperspy/tests/axes/test_data_axis.py#L332-L336

https://github.com/LMSC-NTappy/hyperspy/blob/2e757e940dd11b0d913cb1c78348beb54c133f63/hyperspy/tests/axes/test_data_axis.py#L348-L355

File: hyperspy/axes.py (hunk 11 above)
Comment (LMSC-NTappy) [reply hyperspy/axes.py:564]: It should be pretty harmonised now

File: hyperspy/axes.py (hunk 5 above)
Comment (jlaehne) [reply hyperspy/axes.py]: It is definitely nice to have this optimized, but I agree to your above statement that it does not hurt to override certain methods if they can be implemented more efficiently.

File: hyperspy/axes.py (hunk 10 above)
Comment (jlaehne) [reply hyperspy/axes.py:609]: Weird, now even more warnings have appeared, but one is even for a comment

File: hyperspy/axes.py (hunk 5 above)
Comment (jlaehne) [reply hyperspy/axes.py]: @ecripre I would propose to finish this PR and we can further optimize `value2index` in later PRs, what do you think?

Comment (ericpre) [reply hyperspy/axes.py]: > @ecripre I would propose to finish this PR and we can further optimize `value2index` in later PRs, what do you think?

Yes!

**Comment by jlaehne** [comment]
OK, I merge this PR as the main objectives are more than covered. Further optimization that we discussed at several points can be contributed as new PRs. Thanks again!

**Comment by LMSC-NTappy** [comment]
This was a nice ride. Thank you!
